### ISSUE TITLE
JIT flush can lead to degraded performance in snabbvmx

### ISSUE THREAD

**Issue by mwiget** [Igalia/snabbswitch#1011]
After a JIT flush on long running snabb instances on a production system , we experience a greater likelyhood of more flushes and the breaths/sec show less room after a flush. 

I reproduced this problem on a lab system with 2 snabb instances serving bi-diretional line rate traffic via snabbvmx. Everything is fine during the first 8 hours, just one or two flushes and the breaths/sec stays around 55k before and after during these first 8 hours.Then suddenly one of the instances experienced a flush, followed by many more (47 total) in 5 hours and the remaining breaths/sec stays at 22k after that. 

The setup consists of one server running 2 snabbvmx instances connected to one vMX and 2 separate packetblaster processes firing packets from different PCAP files at line rate over two loopback cables between 2 Intel 82599 NICs (packetblaster processes on the same card).

The graph shows relative timestamps when flushes happen during these 5 hours:
![flushes](https://user-images.githubusercontent.com/3111088/35426029-47b94a4a-0260-11e8-97ae-00ce19836b95.png)
Not shown is the 8 hour flush-free time before the events started. The traffic pattern sent by packetblaster repeats itself a few times per hour, so that can't explain the sudden increase of events.

No flushes experienced on the second snabbvmx process on the same server, serving the same vMX instance via vhostuser (memory mapped). This seems to rule out memory mapping/fragmentation issue as the root cause ASAIK, but open to discussion.

The snabbvmx code in use is this one: 
https://github.com/Juniper/vmx-docker-lwaftr/tree/v1.1.20
https://github.com/mwiget/snabb/tree/1c907062e7721cc75a8183b3232b4da58e34f16c

The JIT flush seems to optimise itself sometimes into a less optimal state, visible via lower breath room. 

The code in use has evolved in the meantime and I'm curious if such behaviour has been seen and if there were changes done how and when the flush triggers.

Possible workaround is a restart of the snabbvmx instance experiencing the drops instead of running jitflush. Once I restarted snabbvmx, its breaths/sec count went back to 55k and stayed there without flushes for at least an hour (until I wrote this issue).

Thoughts?

**Comment by wingo** [comment]
Weird.  Have you applied https://github.com/Igalia/snabb/pull/940 to your code?

**Comment by mwiget** [comment]
No, #940 isn't in there. That fix ignores drops right after a flush for 10 seconds only. Why would this help? Is triggering flush twice (or more) in less than a minute helpful? 
I know we used flush() mainly to survive dramatic changes in workloads, when suddenly traffic changes from 0 to line rate load and back. In a live deployment, changes are less drastic and hence might not need a flush() at all. I'm tempted to change one of the instances to warn only instead of actually running a flush.

**Comment by wingo** [comment]
This would help because a flush will immediately cause drops.  The drops caused by a flush should not cause future flushes.  By the time 10s have passed, JIT compilation has happened already.

RaptorJIT does fix some things related to JIT flushing -- right now some JIT state is still kept in LuaJIT (blacklisted traces).

**Comment by lukego** [comment]
I would love to apply RaptorJIT+Studio to this. Then we should be able to compare the traces before and after the flush to see what significant differences there are.

Would it be possible to repeat this test with the RaptorJIT branch merged? I first need to check and see if any more instrumentation is needed to account for the JIT flushes. The basic idea is to keep trace numbers unique (don't reuse trace numbers after a flush) and log a message when a flush happens to know which traces belong to which "generation." Then the data should be sufficient and we might only need the Studio UI to indicate which generation each trace belongs to. Could also be that we should create new profiler datasets between flushes... anyway I am willing to hack this if you will be able to test it.

**Comment by lukego** [comment]
> RaptorJIT does fix some things related to JIT flushing -- right now some JIT state is still kept in LuaJIT (blacklisted traces).

This was actually ported to RaptorJIT from the Snabb branch of LuaJIT. @alexandergall did we ever push all of that upstream in Snabb?

**Comment by wingo** [comment]
apparently it is upstream: https://github.com/snabbco/snabb/pull/1244

**Comment by mwiget** [comment]
@wingo: I see. I misread the title of the fix and understand now the intent to ignore lost packets during the JIT flush to avoid firing again too quickly. Looking thru my log of the incident in my lab, I do see 8 out of 47 flush'es that fired again after 20 second, but only once, then it was good again until the next incident, which varied from 40 seconds to 1600 seconds until the next flush. 

@lukego: thanks for your offer, maybe I'll take you up on that offer, but need to understand things better first whats happening. (next post)

**Comment by wingo** [comment]
It could certainly be that more JITs result in more blacklists which results in less good JITs, related to the issue fixed in https://github.com/snabbco/snabb/pull/1244

**Comment by mwiget** [comment]
Is a change in breaths/sec an indication that JIT changed its traces and does this happen without a forced flush? I'm running now without flush (set to warn only) and it warned me 3 times and the breaths/s went down from 44k to 22k'ish. Then I restarted snabb again and it went back up to 44k and a single drop warning (right at start).

**Comment by mwiget** [comment]
I applied now #940 and am run them now within my lab for hours to come. Thanks @wingo .

**Comment by mwiget** [comment]
Also applied snabbco#1244 plus a compile error fix, cherry-picked into a branch:
https://github.com/mwiget/snabb/commits/flush-drop-count-fix
I'll run this in my lab, hopefully all weekend.

**Comment by mwiget** [comment]
Good news. I applied #940 and snabbco#1244 plus the GCC 7.1.0 fix (https://github.com/mwiget/snabb/commit/3837c29f539e0ce0b690df825e60546586b84d8f) into branch https://github.com/mwiget/snabb/commits/flush-drop-count-fix and linked to it from 
https://github.com/juniper/vmx-docker-lwaftr, where I tagged version v1.1.21. 

Tests ran all weekend, replaying PCAP files at line rate into 2x10GE ports with vMX and Snabb and only experienced 3 JIT flushes on xe0 and 2 JIT flushes on xe1 while breaths/sec remains healthy between 45k and 65k. The server didn't use any CPU isolation.

I'll close this issue as resolved. Many thanks @wingo and @lukego for helping to solve this!
